### PR TITLE
Add pre-commit.ci

### PR DESCRIPTION
**Pull request by mikemhenry**
We can use this PR to decide what pre-commit hooks we want to use with pre-commit.ci 

https://pre-commit.ci/
https://pre-commit.com/
https://pre-commit.com/hooks.html

I am not too fused on which tools we use but I would like to have things that:

* auto format code
* clean up files (like removing trailing whitespace)
* prevent people from making git mistakes (large files, broken symlinks, etc)

@dotsdl @ianmkenney @IAlibay @atravitz -- Interested in hearing your thoughts!

### CODE REVIEW

**Comment by mikemhenry** [comment]
Related #291 #363

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  - id: check-toml
+  - id: check-yaml
+  - id: debug-statements
+  - id: destroyed-symlinks
+  - id: end-of-file-fixer
+    exclude: '\.(graphml)$'
+  - id: trailing-whitespace
+    exclude: '\.(pdb|gro|top|sdf|xml|cif|graphml)$'
+- repo: https://github.com/psf/black
+  rev: 24.2.0
```
Comment (IAlibay) [inline .pre-commit-config.yaml:20]: I know it's possible with black to choose a "standard", is this the way we do this or does this just control the version of black we install?

Comment (ianmkenney) [reply .pre-commit-config.yaml:20]: I'm also a bit curious here. Which tool is responsible for installing what? For instance, with black, does running the pre-commit install actually install black or just the hook that will execute black?

**Review by ianmkenney — CHANGES_REQUESTED** [review]
I think overall this looks good, but I have a couple of questions/requests

File: .pre-commit-config.yaml (hunk 1 above)
Comment (mikemhenry) [reply .pre-commit-config.yaml:20]: Good question! So if you are just using pre-commit.ci, locally you don't need to do anything. But if you use this locally, `pre-commit` when running for the first time will create a virtual python environment (separate than whatever one you are currently using, this lets you use packages and versions that might not work with your program be used for the pre-commit hooks) and install everything needed for the hooks and run them. When you re-run the hooks, it will re-use the virtual environment it created the first time.

so tl;dr if you install pre-commit, it will create a virtual env that will install all the tools needed, then execute the hooks

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  - id: check-symlinks
+  - id: check-toml
+  - id: check-yaml
+  - id: debug-statements
+  - id: destroyed-symlinks
+  - id: end-of-file-fixer
+    exclude: '\.(graphml)$'
+  - id: trailing-whitespace
+    exclude: '\.(pdb|gro|top|sdf|xml|cif|graphml)$'
+- repo: https://github.com/psf/black
```
[the remark this answers is not in the document]
Comment (mikemhenry) [reply .pre-commit-config.yaml:19]: Good catch!

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  rev: 24.2.0
+  hooks:
+  - id: black
+  - id: black-jupyter
+- repo: https://github.com/PyCQA/isort
+  rev: 5.13.2
+  hooks:
+  - id: isort
+- repo: https://github.com/econchick/interrogate
```
[the remark this answers is not in the document]
Comment (mikemhenry) [reply .pre-commit-config.yaml:28]: The config is controlled in the pyproject.toml (so that way if you run the tool as a standalone thing, you use the same config as the one that runs on pre-commit)

```
[tool.interrogate]
fail-under = 0 
ignore-regex = ["^get$", "^mock_.*", ".*BaseClass.*"]
# possible values for verbose: 0 (minimal output), 1 (-v), 2 (-vv)
verbose = 2 
color = true
exclude = ["build", "docs"]
``` 
Config options are listed here: https://github.com/econchick/interrogate?tab=readme-ov-file#configuration

I don't want the tools to get in our way, so if we find its too annoying and gives us many false positives, we can drop it.

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  rev: 1.5.0
+  hooks:
+    - id: interrogate
+      pass_filenames: false
+- repo: https://github.com/asottile/add-trailing-comma
+  rev: v3.1.0
+  hooks:
+  - id: add-trailing-comma
+- repo: https://github.com/asottile/yesqa
```
[the remark this answers is not in the document]
Comment (mikemhenry) [reply .pre-commit-config.yaml:37]: considering we have 0 `# noqa` comments in `gufe` we can drop it

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+- repo: https://github.com/PyCQA/isort
+  rev: 5.13.2
+  hooks:
+  - id: isort
+- repo: https://github.com/econchick/interrogate
+  rev: 1.5.0
+  hooks:
+    - id: interrogate
+      pass_filenames: false
+- repo: https://github.com/asottile/add-trailing-comma
```
[the remark this answers is not in the document]
Comment (mikemhenry) [reply .pre-commit-config.yaml]: I think there are some cases where black does things differently, but I think dropping it makes sense unless we run into a situation where black isn't putting in a trailing comma where we want it.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (mikemhenry) [reply .pre-commit-config.yaml:20]: Also as a quick follow up, there is a section in the pyproject.toml:

```
[tool.black]
line-length = 120 
```

that controls black's config, so we can set it to use whichever "standard" we want.

**Comment by mikemhenry** [comment]
@IAlibay @ianmkenney - This is ready for another look!
@atravitz - Let me know what you think! I also know you were looking at `ruff` which might be able to replace `black` and `isort` and add stuff like check for unused imports and the like.

**Review by ianmkenney — APPROVED** [review]
This looks good to me! Thanks for putting this together.

**Review by IAlibay — APPROVED** [review]

**Comment by mikemhenry** [comment]
I want to wait for others to provide some feedback, but we can discuss next steps. 

Do we want to just merge this in and fix things as we go on PRs? Or do we want to also run the hooks on the files as part of this PR or a separate PR?

**Comment by ianmkenney** [comment]
@mikemhenry cleaner to run it here once we're happy with the tooling choices

**Comment by dotsdl** [comment]
@mikemhenry please proceed with merge and setup of `pre-commit` CI. After this, we'll want to immediately apply the reformatting to `gufe`.

@atravitz also recommends we tell `git blame` to ignore the commit in which we applied the reformatting in order to maintain the usefulness of `git blame`; see here for details: https://www.stefanjudis.com/today-i-learned/how-to-exclude-commits-from-git-blame/

**Comment by mikemhenry** [comment]
> @mikemhenry cleaner to run it here once we're happy with the tooling choices

> @mikemhenry please proceed with merge and setup of pre-commit CI. After this, we'll want to immediately apply the reformatting to gufe.

@IAlibay @atravitz -- can you break the tie here?

**Comment by ianmkenney** [comment]
> > @mikemhenry cleaner to run it here once we're happy with the tooling choices
> 
> > @mikemhenry please proceed with merge and setup of pre-commit CI. After this, we'll want to immediately apply the reformatting to gufe.
> 
> @IAlibay @atravitz -- can you break the tie here?

@mikemhenry I'll break it---go with David!

**Comment by mikemhenry** [comment]
Cool, I will do the formatting in a separate PR but wait until that is done so there isn't a gap between having .pre-commit going and everything unformatted

**Comment by mikemhenry** [comment]
Also need to add https://pre-commit.ci/#configuration-autofix_prs !!

**Comment by mikemhenry** [comment]
I added a pull request template that mentions how to run the pre-commit.ci

**Comment by mikemhenry** [comment]
Okay so plan is to merge in https://github.com/OpenFreeEnergy/gufe/pull/421 if it is green. Then I will add the commit from that merge to `.git-blame-ignore-revs` which I will add to this PR. That way we will have a functional git history, and this PR will correctly point to "me" as the blame for the config + the ignore recs, which will be helpful for who ever looks at this in the future
